src/time/posix-macos-time.c

```diff
@@ -5,22 +5,42 @@
 
 #include <mach/clock_types.h>
 
-#if __MAC_OS_X_VERSION_MAX_ALLOWED < 101200
+#include <sys/sysctl.h>
+#include <mach/mach_time.h>
+
+// this code is based on https://opensource.apple.com/source/Libc/Libc-1439.141.1/gen/clock_gettime.c.auto.html
+int _clock_gettime_impl(clockid_t clk_id, struct timespec *tp)
+{
+    switch(clk_id){
+    case CLOCK_REALTIME: {
+        struct timeval tv;
+        int ret = gettimeofday(&tv, NULL);
+        TIMEVAL_TO_TIMESPEC(&tv, tp);
+        return ret;
+    }
+    case CLOCK_MONOTONIC: {
+        struct timeval boottime, tv;
+        size_t len = sizeof(tv);
+        int ret = sysctlbyname("kern.boottime", &boottime, &len, NULL, 0);
```

Comment:
The code in the linked Apple page does this:

```c
    case CLOCK_MONOTONIC: {
        struct timeval tv;
        uint64_t boottime_usec;
        int ret = _mach_boottime_usec(&boottime_usec, &tv);
        struct timeval boottime = {
            .tv_sec = boottime_usec / USEC_PER_SEC,
            .tv_usec = boottime_usec % USEC_PER_SEC
        };
        timersub(&tv, &boottime, &tv);
        TIMEVAL_TO_TIMESPEC(&tv, tp);
        return ret;
    }
```

I don't have time to investigate this myself. Could I ask you to explain the lines that are different between Apple and your implementation?

Specifically:

- I understand that the `sysctlbyname("kern.boottime"...` line is supposed to do the same thing as `mach_boottime_usec(&boottime_usec, &tv);` but then
- the lines and conversions are somehow different.

This is just for my understanding but adding a comment that explains the difference with the Apple's implementation could also be useful.

Reply:
@stanislaw let me explain it when!

How you can see original way to `CLOCK_MONOTONIC` is implemented via `_mach_boottime_usec ` which is:
```
static int
_mach_boottime_usec(uint64_t *boottime, struct timeval *realtime)
{
    uint64_t bt1 = 0, bt2 = 0;
    int ret;
    do {
        bt1 = mach_boottime_usec();
        if (os_slowpath(bt1 == 0)) bt1 = _boottime_fallback_usec();

        atomic_thread_fence(memory_order_seq_cst);

        ret = gettimeofday(realtime, NULL);
        if (ret != 0) return ret;

        atomic_thread_fence(memory_order_seq_cst);

        bt2 = mach_boottime_usec();
        if (os_slowpath(bt2 == 0)) bt2 = _boottime_fallback_usec();
    } while (os_slowpath(bt1 != bt2));
    *boottime = bt1;
    return 0;
}
```

This function gets boot time in usec by two possible ways:
 - `mach_boottime_usec()`;
 - _boottime_fallback_usec()`.

I have no idea where `mach_boottime_usec()` is implemented, so, I've simplified my version to use the only `_boottime_fallback_usec()`. This function simple fills provided `struct timeval` with the time that the system was booted. You may read details inside [sysctlbyname man page](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/sysctlbyname.3.html).

`os_slowpath(...)` and `atomic_thread_fence(memory_order_seq_cst)` and the while loop is quite wired here, and I've assumed that it required by unknow for me `mach_boottime_usec()` implementation.

Anyway, the next step this function is getting current time via  `gettimeofday`.

So, `_mach_boottime_usec` fills two `struct timeval`: one with boot time, and the second one with current time. After that simple `timersub` to substract from current time the boot time, allows to get some value which is monotonically increased: `CLOCK_MONOTONIC`.

Reply:
Ah, I guess I understand why they have used `while`.  Probably when a system adjust time from NTP server, it affects also boot time.

And this can create side effect when `CLOCK_MONOTONIC` aren't monotonically increased. It may be decreased after all.

Let me recover this condition.

Reply:
@stanislaw thanks, you raised a good point!